### ISSUE TITLE
Use NTFS hardlinks and junctions instead of NTFS symlinks

### ISSUE THREAD

**Issue by ARF1** [git-for-windows/git#1099]
The new symlink support of Git for Windows is a very welcome addition!

One problem I am having is that for members of the Administrators group, UAC elevation is required for the git executable. This means that the `git.exe` properties must be set as `Run as administrator`. The results is that for EVERY SINGLE git command, a UAC prompt is displayed.

An equivalent solution that could be implemented without any need for UAC elevation using NTFS hardlinks (for files) and NTFS junctions (for directories). In contrast to NTFS symlinks, these can be created without UAC elevation.

This would be much more user friendly.

### Setup

 - Git for Windows 32-bit

```
$ git --version --build-options

git version 2.12.0.windows.1
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 6.3.9600]
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git clone -c core.symlinks=true <URL>
```
 - What did you expect to occur after running these commands?

Clone of repo without UAC prompt

 - What actually happened instead?

Command fails for users in the Adminstrators group. After setting git.exe properties "Run as administrator" a UAC prompt is displayed for every single git command.

**Comment by whoisj** [comment]
@ARF1 please see https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10/#iUAXAL3HmoCPjRtr.97 which discusses symlinks on Windows 10. Recently, symlinks in Windows no longer require elevated privileges as long as you have a new enough version of Windows and Developer Mode activated.

**Comment by ARF1** [comment]
@whoisj Indeed. Is there any reason not to make it work seamlessly on previous Windows versions as well?

**Comment by whoisj** [comment]
> Is there any reason not to make it work seamlessly on previous Windows versions as well?

It's not specific to versions of Windows only. The underlying file system needs to support them as well. Hard links are limited to NTFS only, whereas symbolic links are supported by NTFS, ReFS, and SMB. I do not believe FAT gets any love.

Another drawback to hard links on Windows, is that only the updated link gets its mtime and size data updated. This would be rather annoying / difficult / impossible for Git to handle correctly.

From [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365006(v=vs.85).aspx):
>  Any changes to that file are instantly visible to applications that access it through the hard links that reference it. However, the directory entry size and attribute information is updated only for the link through which the change was made.

The downside to soft..., er "symbolic" links is that if the target file is deleted the links become inoperative. Unlike hard links, there's no reference counting. I, personally, prefer hard links but also hate the fact that meta data isn't shared.

Basically, linking on Windows is broken and there's no resolution in the foreseeable future.

**Comment by ARF1** [comment]
> It's not specific to versions of Windows only. The underlying file system needs to support them as well. Hard links are limited to NTFS only, whereas symbolic links are supported by NTFS, ReFS, and SMB.

@whoisj  I did not know that. It is a very convincing reason for sticking with symlinks. Thank you for enlightening me.

Since hard links are definitely ruled out, would it be possible to make Git for Windows more useable for users on previous windows versions?

At present, on many Windows versions, creation on symlinks requires UAC elevation but git itself does not request it. This means, that on these versions, one has to trigger UAC elevation for every execution of git (even if UAC elevation is not required for the command) which makes it pretty much unuseable.

Is there any way for Git to show a UAC prompt if and only if a command requires creating symlinks?

**Comment by dscho** [comment]
Neither hardlinks nor junctions are equivalent to symbolic links. Please see https://github.com/git-for-windows/git/wiki/Symbolic-links for more information about symbolic links.
